### PR TITLE
Return of Chemical Macros

### PR DESCRIPTION
**Pull request by karmaisblackandbluepilled**
Recording sucks balls and was added by or*nges, macros are better.
I need someone with TGUI skills to tell me why recipes still show up as this.
image

## Changelog
:cl:
add: Added the Chemical Macros back to chem dispensers. God bless powergamers.
/:cl:

### CODE REVIEW

**Comment by eeSPee** [comment]
My guess is the recipe requires 0.45u of solution and it rounds down to 0.

File: tgui/packages/tgui/interfaces/ChemDispenser.js [hunk 1]
```diff
@@ -111,7 +111,7 @@ export const ChemDispenser = (props, context) => {
                 lineHeight={1.75}
                 content={recipe.name}
```
Comment (eeSPee) [inline tgui/packages/tgui/interfaces/ChemDispenser.js:112]: ...but accessed as recipe.name here...

File: code/modules/reagents/chemistry/machinery/chem_dispenser.dm [hunk 2]
```diff
@@ -288,73 +287,92 @@
...
-				recording_recipe = null
-				. = TRUE
-		if("cancel_recording")
-			if(!is_operational)
-				return
-			recording_recipe = null
-			. = TRUE
+				if (resmismatch && alert("[src] is not yet capable of replicating this recipe with the precision it needs, do you want to save it anyway?",, "Yes","No") == "No")
+					return
+				saved_recipes += list(list("recipe_name" = name, "contents" = recipe))
```
[the remark this answers is not in the document]
Comment (karmaisblackandbluepilled) [reply code/modules/reagents/chemistry/machinery/chem_dispenser.dm:346]: Yeah, but if I do recipe_name in the .js, it makes the box empty entirely. That's the problem, I don't know why.

[the remark this answers is not in the document]
Comment (eeSPee) [reply code/modules/reagents/chemistry/machinery/chem_dispenser.dm:346]: Would it fix the issue if you save it as "name"?

[the remark this answers is not in the document]
Comment (eeSPee) [reply code/modules/reagents/chemistry/machinery/chem_dispenser.dm:346]: Or try recipe.recipe_name

[the remark this answers is not in the document]
Comment (karmaisblackandbluepilled) [reply code/modules/reagents/chemistry/machinery/chem_dispenser.dm:346]: as I said, recipe.recipe_name makes it empty

[the remark this answers is not in the document]
Comment (eeSPee) [reply code/modules/reagents/chemistry/machinery/chem_dispenser.dm:346]: Either replace recipe_name to simply name, or replace all instances of recipe.name to recipe.recipe_name

[the remark this answers is not in the document]
Comment (karmaisblackandbluepilled) [reply code/modules/reagents/chemistry/machinery/chem_dispenser.dm:346]: Nope.

[the remark this answers is not in the document]
Comment (eeSPee) [reply code/modules/reagents/chemistry/machinery/chem_dispenser.dm:346]: Why does it add list(list(...)) instead of just list(...)? Could this be the problem?

File: code/modules/reagents/chemistry/machinery/chem_dispenser.dm [hunk 3]
```diff
@@ -288,73 +287,92 @@
...
+						if (!resmismatch && !check_macro_part(reagents))
+							resmismatch = TRUE
+						continue
+					else
+						var/chemid = reagent[1]
+						visible_message("<span class='warning'>[src] buzzes.</span>", "<span class='italics'>You hear a faint buzz.</span>")
+						to_chat(usr, "<span class='danger'>[src] cannot find Chemical ID: <b>[chemid]</b>!</span>")
+						playsound(src, 'sound/machines/buzz-two.ogg', 50, 1)
 						return
-				saved_recipes[name] = recording_recipe
```
Comment (EtraIV) [inline code/modules/reagents/chemistry/machinery/chem_dispenser.dm:346]: `saved_recipes[name] = recording_recipe` to `saved_recipes += list(list("recipe_name" = name, "contents" = recipe))`, This  is probably what's breaking it. Instead of adding a recipe value with the key of the recipe's name, you're adding a list that contains the list that contains the recipe name and the recipe iteslf as values. This is triply nested lists. `saved_recipes[name] = recipe` should ail your issues.

File: code/modules/reagents/chemistry/machinery/chem_dispenser.dm [hunk 4]
```diff
@@ -222,6 +222,7 @@
 		data["beakerCurrentpH"] = null
 
 	var/chemicals[0]
+	var/recipes[0]
```
Comment (EtraIV) [inline code/modules/reagents/chemistry/machinery/chem_dispenser.dm:225]: You turned the recipes associative list, which used the keys of the recipes as the recipe names, into a numbered list, which still uses the keys (now numbers, instead of the recipe name) as its name. Good job.

File: code/modules/reagents/chemistry/machinery/chem_dispenser.dm [hunk 5]
```diff
@@ -233,9 +234,10 @@
 				chemname = "[pick_list_replacements("hallucination.json", "chemicals")]"
 			chemicals.Add(list(list("title" = chemname, "id" = ckey(temp.name), "pH" = temp.ph, "pHCol" = convert_ph_to_readable_color(temp.ph))))
 	data["chemicals"] = chemicals
-	data["recipes"] = saved_recipes
+	for(var/recipe in saved_recipes)
+		recipes.Add(list(recipe))
```
Comment (EtraIV) [inline code/modules/reagents/chemistry/machinery/chem_dispenser.dm:238]: What the fuck?

**Review by EtraIV — CHANGES_REQUESTED** [review]
You fucking broke this yourself by turning the list of chemical recipes from an associative list into a numbered one. Undo that shit and it should be fine.

File: code/modules/reagents/chemistry/machinery/chem_dispenser.dm [hunk 6]
```diff
@@ -288,73 +287,92 @@
 			replace_beaker(usr)
 			. = TRUE
 		if("dispense_recipe")
-			if(!is_operational || QDELETED(cell))
```
Comment (EtraIV) [inline code/modules/reagents/chemistry/machinery/chem_dispenser.dm:291]: Why did you remove this?

File: code/modules/reagents/chemistry/machinery/chem_dispenser.dm [hunk 7]
```diff
@@ -288,73 +287,92 @@
...
-				return
-
-			var/list/chemicals_to_dispense = saved_recipes[params["recipe"]]
-			if(!LAZYLEN(chemicals_to_dispense))
```
Comment (EtraIV) [inline code/modules/reagents/chemistry/machinery/chem_dispenser.dm:295]: Why did you remove this?

File: code/modules/reagents/chemistry/machinery/chem_dispenser.dm [hunk 8]
```diff
@@ -288,73 +287,92 @@
...
+			for(var/key in chemicals_to_dispense) // i suppose you could edit the list locally before passing it
+				var/list/keysplit = splittext(key," ")
+				var/r_id = GLOB.name2reagent[translate_legacy_chem_id(keysplit[1])]
+				if(beaker && dispensable_reagents.Find(r_id)) // but since we verify we have the reagent, it'll be fine
+					var/datum/reagents/R = beaker.reagents
+					var/free = R.maximum_volume - R.total_volume
+					var/rounded = is_resolution(chemicals_to_dispense[key]) ? chemicals_to_dispense[key] : round(chemicals_to_dispense[key], macroresolution)
+					var/actual = min(rounded, (cell.charge * powerefficiency)*10, free)
+					if(actual)
+						if(!cell.use(abs(actual) / powerefficiency))
```
Comment (EtraIV) [inline code/modules/reagents/chemistry/machinery/chem_dispenser.dm:303]: Why did you remove the null check? This can cause runtimes, especially since you're no longer checking if there's a cell in the machine.

File: code/modules/reagents/chemistry/machinery/chem_dispenser.dm [hunk 9]
```diff
@@ -288,73 +287,92 @@
...
+		if("record_recipe") //I'm not renaming this lol
 			if(!is_operational)
 				return
-			recording_recipe = list()
-			. = TRUE
-		if("save_recording")
-			if(!is_operational)
-				return
-			var/name = stripped_input(usr,"Name","What do you want to name this recipe?", "Recipe", MAX_NAME_LEN)
+			var/name = stripped_input(usr,"Name","What do you want to name this recipe?(This is broken)", "Recipe", MAX_NAME_LEN)
```
Comment (EtraIV) [inline code/modules/reagents/chemistry/machinery/chem_dispenser.dm:321]: Remove this comment.

**Review by EtraIV — CHANGES_REQUESTED** [review]
Go home and die.

File: code/modules/reagents/chemistry/machinery/chem_dispenser.dm (hunk 6 above)
Comment (karmaisblackandbluepilled) [reply code/modules/reagents/chemistry/machinery/chem_dispenser.dm:291]: I don't know why but QDELETED(CELL) caused it to runtime. It still has the !is_oeprational.

File: code/modules/reagents/chemistry/machinery/chem_dispenser.dm (hunk 7 above)
Comment (karmaisblackandbluepilled) [reply code/modules/reagents/chemistry/machinery/chem_dispenser.dm:295]: remove what

File: code/modules/reagents/chemistry/machinery/chem_dispenser.dm (hunk 8 above)
Comment (karmaisblackandbluepilled) [reply code/modules/reagents/chemistry/machinery/chem_dispenser.dm:303]: I wish github actually pointed to what you're talking about because I can't see what null check you're pointing to. The ? in the !cell?.
Probably a mistake. I wil add it back.

File: code/modules/reagents/chemistry/machinery/chem_dispenser.dm (hunk 7 above)
Comment (karmaisblackandbluepilled) [reply code/modules/reagents/chemistry/machinery/chem_dispenser.dm:295]: You mean if(!LAZYLEN(chemicals_to_dispense))? Mainly because it doesn't seem to actually do anything?

File: code/modules/reagents/chemistry/machinery/chem_dispenser.dm (hunk 2 above)
[the remark this answers is not in the document]
Comment (EtraIV) [reply code/modules/reagents/chemistry/machinery/chem_dispenser.dm:346]: This is not resolved. Change it back to what it was (`saved_recipes[name] = recording_recipe`)

File: code/modules/reagents/chemistry/machinery/chem_dispenser.dm [hunk 10]
```diff
@@ -288,73 +287,92 @@
...
+		if(amt % i == 0)
+			return TRUE
+
+/obj/machinery/chem_dispenser/proc/process_recipe_list(list/recipe)
+	var/list/key_list = list()
+	var/list/final_list = list()
+	var/list/first_process = splittext(recipe["contents"], ";")
+	for(var/reagents in first_process)
+		var/list/splitreagent = splittext(reagents, "=")
+		final_list += list(avoid_assoc_duplicate_keys(splitreagent[1],key_list) = text2num(splitreagent[2]))
```
Comment (EtraIV) [inline code/modules/reagents/chemistry/machinery/chem_dispenser.dm:373]: `final_list[avoid_assoc_duplicate_keys(splitreagent[1], key_list)] = text2num(splitreagent[2])`
You REALLY don't know how to use assoc lists.

**Review by EtraIV — CHANGES_REQUESTED** [review]

File: tgui/packages/tgui/interfaces/ChemDispenser.js [hunk 11]
```diff
@@ -111,7 +111,7 @@ export const ChemDispenser = (props, context) => {
...
                 onClick={() => act('dispense_recipe', {
-                  recipe: recipe.name,
+                  recipe: recipe.contents,
                 })} />
```
Comment (EtraIV) [inline tgui/packages/tgui/interfaces/ChemDispenser.js:115]: Change this back.

File: code/modules/reagents/chemistry/machinery/chem_dispenser.dm (hunk 9 above)
Comment (EtraIV) [reply code/modules/reagents/chemistry/machinery/chem_dispenser.dm:321]: It's not fucking resolved you dickweed, REMOVE THIS COMMENT

File: code/modules/reagents/chemistry/machinery/chem_dispenser.dm (hunk 6 above)
Comment (EtraIV) [reply code/modules/reagents/chemistry/machinery/chem_dispenser.dm:291]: Show the runtime, this shouldn't be possible since it runs a check on whether or not the cell actually exists at all.

File: code/modules/atmospherics/machinery/datum_pipeline.dm [hunk 12]
```diff
@@ -149,6 +149,8 @@
 /datum/pipeline/proc/merge(datum/pipeline/parent_pipeline)
 	if(parent_pipeline == src)
 		return
+	if(!parent_pipeline)
```
Comment (EtraIV) [inline code/modules/atmospherics/machinery/datum_pipeline.dm:152]: Why are you adding this again? This got through in the forge PR. Also doing `if(!parent_pipeline || parent_pipeline == src)` would've been a lot better.

File: code/modules/reagents/chemistry/machinery/chem_dispenser.dm (hunk 9 above)
Comment (karmaisblackandbluepilled) [reply code/modules/reagents/chemistry/machinery/chem_dispenser.dm:321]: clueless

File: code/modules/atmospherics/machinery/datum_pipeline.dm (hunk 12 above)
Comment (karmaisblackandbluepilled) [reply code/modules/atmospherics/machinery/datum_pipeline.dm:152]: I forgot to remove it. I added it because it was annoying me!

File: code/modules/reagents/chemistry/machinery/chem_dispenser.dm (hunk 10 above)
Comment (karmaisblackandbluepilled) [reply code/modules/reagents/chemistry/machinery/chem_dispenser.dm:373]: yeah, I already said this

Comment (karmaisblackandbluepilled) [reply code/modules/reagents/chemistry/machinery/chem_dispenser.dm]: etra you coon there's no `recording_recipe` in the whole code anymore

File: tgui/packages/tgui/interfaces/ChemDispenser.js (hunk 11 above)
Comment (karmaisblackandbluepilled) [reply tgui/packages/tgui/interfaces/ChemDispenser.js]: you are retarded!

File: code/modules/reagents/chemistry/machinery/chem_dispenser.dm (hunk 2 above)
[the remark this answers is not in the document]
Comment (karmaisblackandbluepilled) [reply code/modules/reagents/chemistry/machinery/chem_dispenser.dm]: etra you coon there's no recording_recipe in the whole code anymore

File: tgui/packages/tgui/interfaces/ChemDispenser.js (hunk 11 above)
Comment (EtraIV) [reply tgui/packages/tgui/interfaces/ChemDispenser.js:115]: i was under the assumption that the process_recipe_list proc would directly read from the recipe list variable, rather than directly parse it from the parameter, and it seems this difference varies across PRs.

**Review by EtraIV — APPROVED** [review]
